### ISSUE TITLE
2 active directions when D-Pad quickly changes direction

### ISSUE THREAD

**Issue by ImAReplicant** [godotengine/godot#42200]
**Godot version:**

Godot 3.2.3 and 3.2.2

**OS/device including version:**

Arch Linux

**Issue description:**

Change quickly the direction on D-Pad of gamepad doesn't released the first direction.
Tested with Xbox Controller, Dualshock 4 and Nintendo Switch Controller

**Steps to reproduce:**
Download the "Joypads Demo / Tools" and quickly switch from left to right or top to bottom for example, both directions remain pressed :
![gamepad](https://user-images.githubusercontent.com/71564949/93687811-fa67e880-fac0-11ea-9892-ff6d667ef5d5.png)

**Comment by quinnyo** [comment]
I just tried this with a PS4 controller but couldn't replicate it. Maybe I just can't switch directions as fast as I used to... 
I thought it might be worth mentioning that I have had some pretty surprising behaviour with controllers before, when Steam is running. Something to do with the steam input customisation fighting the system input drivers, perhaps.

(on a reasonably up to date Arch Linux, Godot 3.2.3, bluetooth controller)

**Comment by ImAReplicant** [comment]
Yeah on a PS4 controller it's harder to reproduce but after several tries the bug appears anyway.
On a switch pro controller or xbox 360 controller, it's much easier, probably due to the design of the D-Pad :

![ps4](https://user-images.githubusercontent.com/71564949/93711705-b58b9280-fb50-11ea-8fca-b2ff9c285efd.gif)
![switch](https://user-images.githubusercontent.com/71564949/93711754-174bfc80-fb51-11ea-998d-5ce6e2dbe2b0.gif)

For Steam I don't know if it's the problem but I have the same bug with or without Steam open.

**Comment by Calinou** [comment]
I can't reproduce this on Godot 3.2.3 (Fedora 31, Xbox One V3 controller in wired mode).

**Comment by ImAReplicant** [comment]
I tested on Windows with Xbox 360 and PS4 controllers and I can't reproduce the bug.
I could not test with the Nintendo Switch controller because there is no Windows driver.

So I'm lost, it would be a problem with Linux drivers ?

**Comment by Calinou** [comment]
I wonder if the bug is due to input accumulation…

@ImAReplicant Try calling `Input.set_use_accumulated_input(false)` in the `_ready()` function of any script of the Joypads demo and see if the situation improves.

**Comment by ImAReplicant** [comment]
Always the same bug after the modification.

**Comment by quinnyo** [comment]
I'm not sure how Godot reads/receives gamepad input but AFAIK, via evdev or /dev/input/js*, a D-Pad (hat switch) cannot report opposite directions simultaneously. There's 2 values, for X & Y, essentially a 2 axis joystick. You can run test utilities like evtest or jstest to see this.

**Comment by Calinou** [comment]
It might be worth testing with udev as well. If you can [compile Godot from source](https://docs.godotengine.org/en/latest/development/compiling/compiling_for_linuxbsd.html), pass the `udev=yes` SCons option when building the editor and check again.

**Comment by ImAReplicant** [comment]
@quinnyo Indeed I tested with the 2 utilities and the D-pad is recognized as an axis (ABS_HAT0X and ABS_HAT0Y)

@Calinou Okay, I'll try that.

**Comment by Hex4F** [comment]
I get this same problem on Ubuntu 20.10 with the Godot 3.2.3 binary release. I'm using an Xbox 360 controller and don't have Steam installed. It happens frequently enough to make challenging segments unplayable with the D-pad.

Notably, this never happened when I used Godot on Windows 7 with the exact same hardware.

**Comment by akien-mga** [comment]
> By default, Godot is built without udev support as its presence isn't guaranteed on all Linux distributions.

FYI, official Godot binaries have always been built with udev enabled. Custom builds weren't until a few months ago.

**Comment by mrbobs** [comment]
Hi, I may have run in to the same issue as the one described, is there a fix or workaround for this issue?

**Comment by Rindbee** [comment]
It may happen on specific handles, similar to #66878.

If it is convenient, you can use `evtest`/`evemu-record` to test the keycode mapping of the controller on Linux.

**Comment by DeeJayLSP** [comment]
On my end it happens with a Xbox 360 controller, but not with combined Joy-Cons

**Comment by Calinou** [comment]
@DeeJayLSP Can you reproduce this on the Xbox 360 controller on https://hardwaretester.com/gamepad or other games/programs that let you see gamepad inputs?

Note that in general, games should aim to handle situations where opposite cardinal directions (SOCD) are inputted by the player. Not only this can be due to a defective controller, a software bug of some kind, but also analog keyboards that emulate controllers or macros.

This is typically done by having both inputs cancel each other out (what Godot does with `Input.get_axis()` and `Input.get_vector()`), prioritize the last input, or always prioritize one input (what a lot of games do through their `if`/`else if` statements).

Therefore, it may just be better to document this instead of filtering out events on our end.

**Comment by Rindbee** [comment]
> The code of today correctly clears the opposite direction flag:

It converts a joypad_event into two button events. And it is not processed in the order of release first and then press.

https://github.com/godotengine/godot/blob/78c9f8ddd906273a5970518d54d2f5114d07b246/core/input/input.cpp#L1348-L1357

**Comment by quinnyo** [comment]
> It converts a joypad_event into two button events. And it is not sent in the order of release first and then press.

That seems like a separate issue? -- for one thing, the joypad demo doesn't use input events.

**Comment by DeeJayLSP** [comment]
> [@DeeJayLSP](https://github.com/DeeJayLSP?rgh-link-date=2025-03-10T15%3A56%3A45.000Z) Can you reproduce this on the Xbox 360 controller on [hardwaretester.com/gamepad](https://hardwaretester.com/gamepad) or other games/programs that let you see gamepad inputs?

hardwaretester doesn't provide enough info. But I can use the same controller in other applications, the issue only happens if it's Godot based.

**Comment by Rindbee** [comment]
@DeeJayLSP You can use `evtest`/`evemu-record` to test how many events are emitted by a key press, some devices will emit two, for example https://github.com/godotengine/godot/issues/66878#issuecomment-2231491673.

The event mentioned in https://github.com/godotengine/godot/issues/42200#issuecomment-2714414966 is just a small probability event.

**Comment by DeeJayLSP** [comment]
> [@DeeJayLSP](https://github.com/DeeJayLSP?rgh-link-date=2025-03-12T00%3A02%3A15.000Z) You can use `evtest`/`evemu-record` to test how many events are emitted by a key press, some devices will emit two, for example [#66878 (comment)](https://github.com/godotengine/godot/issues/66878#issuecomment-2231491673).
```
Properties:
Testing ... (interrupt to exit)
Event: time 1741738040.657610, type 1 (EV_KEY), code 311 (BTN_TR), value 1
Event: time 1741738040.657610, -------------- SYN_REPORT ------------
Event: time 1741738041.129628, type 1 (EV_KEY), code 311 (BTN_TR), value 0
Event: time 1741738041.129628, -------------- SYN_REPORT ------------
Event: time 1741738043.631678, type 1 (EV_KEY), code 705 (BTN_TRIGGER_HAPPY2), value 1
Event: time 1741738043.631678, type 3 (EV_ABS), code 16 (ABS_HAT0X), value 1
Event: time 1741738043.631678, -------------- SYN_REPORT ------------
Event: time 1741738044.679622, type 1 (EV_KEY), code 705 (BTN_TRIGGER_HAPPY2), value 0
Event: time 1741738044.679622, type 3 (EV_ABS), code 16 (ABS_HAT0X), value 0
Event: time 1741738044.679622, -------------- SYN_REPORT ------------
```

The first key was a trigger press/release, but the other was d-pad

**Comment by Rindbee** [comment]
> The first key was a trigger press/release, but the other was d-pad

So your issue is a duplicate of https://github.com/godotengine/godot/issues/66878#issuecomment-2231491673.

**Comment by Nintorch** [comment]
Hello! May I ask if the issue is still reproducible in Godot 4.5 and 4.6?
